README.md
[![Travis Status][trav_img]][trav_site]
[![AppVeyor Status][appveyor_img]][appveyor_site]
[![Coverage Status][cov_img]][cov_site]
[![NPM Package][npm_img]][npm_site]
[![Dependency Status][david_img]][david_site]
![gzipped size][size_img]
[![Maintenance Status][maintenance-image]](#maintenance-status)

# Radium

```sh
yarn add radium
# or
npm install --save radium
```

Radium is a set of tools to manage inline styles on React elements. It gives you powerful styling capabilities without CSS.

_Inspired by_ <a href="https://speakerdeck.com/vjeux/react-css-in-js">React: CSS in JS</a>
by <a href="https://twitter.com/Vjeux">vjeux</a>.

## Maintenance Status

**Archived:** This project is no longer maintained by Formidable. We are no longer responding to issues or pull requests unless they relate to security concerns. We encourage interested developers to fork this project and make it their own!

## Overview

Eliminating CSS in favor of inline styles that are computed on the fly is a powerful approach, providing a number of benefits over traditional CSS:

- Scoped styles without selectors
- Avoids specificity conflicts
- Source order independence
- Dead code elimination
- Highly expressive

Despite that, there are some common CSS features and techniques that inline styles don't easily accommodate: media queries, browser states (:hover, :focus, :active) and modifiers (no more .btn-primary!). Radium offers a standard interface and abstractions for dealing with these problems.

When we say expressive, we mean it: math, concatenation, regex, conditionals, functions–JavaScript is at your disposal. Modern web applications demand that the display changes when data changes, and Radium is here to help.

For a short technical explanation, see [How does Radium work?](#how-does-radium-work).

## Features

- Conceptually simple extension of normal inline styles
- Browser state styles to support `:hover`, `:focus`, and `:active`
- Media queries
- Automatic vendor prefixing
- Keyframes animation helper
- ES6 class and `createClass` support

## Docs

- [Overview][docs_guides]
- [API Docs][docs_api]
- [Frequently Asked Questions (FAQ)][docs_faq]

## Usage

Start by wrapping your component class with `Radium()`, like `export default Radium(Component)`, or `Component = Radium(Component)`, which works with classes, `createClass`, and stateless components (functions that take props and return a ReactElement). Then, write a style object as you normally would with inline styles, and add in styles for interactive states and media queries. Pass the style object to your component via `style={...}` and let Radium do the rest!

```jsx
<Button kind="primary">Radium Button</Button>
```

```jsx
import Radium from 'radium';
import React from 'react';
import color from 'color';

class Button extends React.Component {
  static propTypes = {
    kind: PropTypes.oneOf(['primary', 'warning']).isRequired
  };

  render() {
    // Radium extends the style attribute to accept an array. It will merge
    // the styles in order. We use this feature here to apply the primary
    // or warning styles depending on the value of the `kind` prop. Since its
    // all just JavaScript, you can use whatever logic you want to decide which
    // styles are applied (props, state, context, etc).
    return (
      <button style={[styles.base, styles[this.props.kind]]}>
        {this.props.children}
      </button>
    );
  }
}

Button = Radium(Button);

// You can create your style objects dynamically or share them for
// every instance of the component.
var styles = {
  base: {
    color: '#fff',

    // Adding interactive state couldn't be easier! Add a special key to your
    // style object (:hover, :focus, :active, or @media) with the additional rules.
    ':hover': {
      background: color('#0074d9')
        .lighten(0.2)
        .hexString()
    }
  },

  primary: {
    background: '#0074D9'
  },

  warning: {
    background: '#FF4136'
  }
};
```

## Importing Radium

As of `v0.22.x`, Radium is built as an ECMAScript Modules-first project. We now have a `package.json:module` entry pointing to our library files with `import|export` statements instead of CommonJS `require`s. We still support CommonJS `require`s with a special `package.json:main` entry pointing to root `index.js` to smooth over this transition. The basic takeaways are:

If you are using **ESM** with **webpack** or **`@std/esm`** with **Node.js**, imports like the following work fine without any gotchas:

```js
import Radium from 'radium';
import Radium, {Style} from 'radium';
```

If you are using **CommonJS** with **Node.js** or **webpack@1** requires work like normal:

```js
const Radium = require('radium');
const {Style} = require('radium');
```

If you are using **CommonJS** with **webpack@2+**, however, you must instead add `.default` to the root `Radium` object import:

```js
const Radium = require('radium').default; // CHANGED: Must add `.default`
const {Style} = require('radium'); // Works as per normal
```

If you cannot change the `require` statements directly (say Radium is included from a different library your project depends on) you can manually tweak the Radium import in your project's webpack configuration with the following:

```js
resolve: {
  alias: {
    radium: require.resolve('radium/index');
  }
}
```

which will allow `const Radium = require('radium');` to still work. The configuration effectively forces webpack to point to code from `package.json:main` (which points to `/index.js`) instead of what is in `package.json:module`.

_Note:_ Radium uses `Reflect` which is not supported in IE11. You will need to bring in a polyfill like [CoreJs](https://github.com/zloirock/core-js#ecmascript-reflect) in order to support <IE11.

## Examples

To see the universal examples:

```
npm install
npm run universal
```

To see local client-side only examples in action, do this:

```
npm install
npm run examples
```

## How does Radium work?

Following is a short technical explanation of Radium's inner workings:

- Wrap the `render` function
- Recurse into the result of the original `render`
- For each element:
  - Add handlers to props if interactive styles are specified, e.g. `onMouseEnter` for `:hover`, wrapping existing handlers if necessary
  - If any of the handlers are triggered, e.g. by hovering, Radium calls `setState` to update a Radium-specific field on the components state object
  - On re-render, resolve any interactive styles that apply, e.g. `:hover`, by looking up the element's key or ref in the Radium-specific state

## More with Radium

You can find a list of other tools, components, and frameworks to help you build with Radium on our [wiki](https://github.com/FormidableLabs/radium/wiki). Contributions welcome!

## Contributing

Please see [CONTRIBUTING](https://github.com/FormidableLabs/radium/blob/master/CONTRIBUTING.md)

[trav_img]: https://api.travis-ci.com/FormidableLabs/radium.svg
[trav_site]: https://travis-ci.com/FormidableLabs/radium
[cov_img]: https://img.shields.io/coveralls/FormidableLabs/radium.svg
[cov_site]: https://coveralls.io/r/FormidableLabs/radium
[npm_img]: https://img.shields.io/npm/v/radium.svg
[npm_site]: https://www.npmjs.org/package/radium
[david_img]: https://img.shields.io/david/FormidableLabs/radium.svg
[david_site]: https://david-dm.org/FormidableLabs/radium
[size_img]: https://badges.herokuapp.com/size/npm/radium/dist/radium.min.js?gzip=true&label=gzipped
[docs_guides]: https://github.com/FormidableLabs/radium/tree/master/docs/guides
[docs_api]: https://github.com/FormidableLabs/radium/tree/master/docs/api
[docs_faq]: https://github.com/FormidableLabs/radium/tree/master/docs/faq
[appveyor_img]: https://ci.appveyor.com/api/projects/status/github/formidablelabs/radium?branch=master&svg=true
[appveyor_site]: https://ci.appveyor.com/project/ryan-roemer/radium
[maintenance-image]: https://img.shields.io/badge/maintenance-stable-blue.svg


CHANGELOG.md
# Radium Changelog

## 0.26.2

- Update `peerDependencies` for React to include 17.

## 0.26.1 (September 29, 2020)

#### Dependency Upgrades

- Bump handlebars from 4.1.2 to 4.5.3 (#1035)
- Bump elliptic from 6.4.0 to 6.5.3 (#1045)
- Bump eslint-utils from 1.3.1 to 1.4.3 (#1039)
- Bump tree-kill from 1.2.0 to 1.2.2 (#1046)
- Bump websocket-extensions from 0.1.1 to 0.1.4 (#1044)
- Bump fstream from 1.0.11 to 1.0.12 (#1038)
- Bump sshpk from 1.11.0 to 1.16.1 (#1037)
- Bump webpack-dev-server from 2.11.1 to 3.1.11 (#1036)
- Bump is-my-json-valid from 2.16.0 to 2.20.5 (#1048)
- Bump http-proxy from 1.16.2 to 1.18.1 (#1047)

#### Bug Fixes

- Fix initial enhancer state for fn components (#1041)

## 0.26.0 (September 5, 2019)

#### Bug Fixes

- Add hook and forwardRef support (#1027)

#### Breaking Changes

- Radium now requires React@16.8.0+ (#1027)

## 0.25.2 (June 7, 2019)

#### Bug Fixes

- Bump handlebars from 4.0.6 to 4.1.2 (#1026)
- Fixed bug with undefined \_extraRadiumStateKeys (#1025)
- Update unitless CSS property list and licensing (#1020)

## 0.25.1 (December 16, 2018)

#### Bug Fixes

- Pass `snapshot` argument to `componentDidUpdate (#1013)

#### Infra/Tooling

- Fix console warnings when running examples (#1002)
- Upgrade karma to 3.0 (#1003)
- Refactor enhancer for readability (#1004)
- Replace Isparta with Istanbul (#1011)
- Prettier (#1012)

## 0.25.0 (September 16, 2018)

- Use `Reflect` to construct es native classes in a way that preserves the value of `this` used in
  the constructor. (#999, #1001)
- Handle multiple animationName props (#909)

## 0.24.1 (July 9, 2018)

- Make `<StyleSheet>` manually update `<style>` tag DOM, fixing media queries on component level flickering. (#626, #950)

## 0.24.0 (March 29, 2018)

- Fix `render` methods that return array of children or `React.Fragment`. (#950)
- Upgrade tests to React 16.2 (for `Fragment` support)

## 0.23.0 (March 15, 2018)

- Support ES7 arrow functions for React class methods. (#738)

## 0.22.1 (March 1, 2018)

- Fix `keyframes` bug from prefixed inline styles. (#973)

## 0.22.0 (February 9, 2018)

### Breaking Changes

- Radium now exports defaults as `.default`, so for runtimes like Node.js for all files in `lib/**`. We have changed `package.json:main` to point to `/index.js` instead of `/lib/index.js` as a convenience wrapper to expose mostly what was there before so behavior of `const Radium = require('radium');` works mostly as it did before. Caveats:
  - When using webpack2+ to build code with `require('radium')` in it you will need to change that to become `require('radium').default`.
  - Any imports of a default export from a file in lib like `const Enhancer = require('radium/lib/enhancer');` will need to be changed to `const Enhancer = require('radium/lib/enhancer').default;`.
  - We have a full examples repository of how imports work in all likely scenarios that should come up. https://github.com/FormidableLabs/radium-experiments-v0.22

### Features

- Add `es` ESM module export files.

### Fixes

- Fix `package.json:scripts.postinstall` task to correctly work for git-based dependencies.

## 0.21.2 (January 25, 2018)

- Fix multiple-value prefixed inline styles. (#962, #958, #951)

## 0.21.1 (January 18, 2018)

- Call `componentDidUpdate()` inherited method (#957).

## 0.21.0 (January 8, 2018)

- Automatically clear browser state of elements when unmounting and remounting (#956).
  - `resolveStyles` returns `{ extraRadiumStateKeys, element }` instead of just `element`.

## 0.20.1 (January 8, 2018)

- Fix `v0.20.0` build.

## 0.20.0 (January 8, 2018)

- Upgrade `inline-style-prefixer` to version `^4.0.0`.

## 0.19.6 (October 5, 2017)

- Allow native ES classes to be used with the Radium enhancer.

## 0.19.5 (October 5, 2017)

- Enable React 16 in `peerDependencies` and update tests to React 16.

## 0.19.4 (July 21, 2017)

- Remove `.babelrc` from published npm registry package.

## 0.19.3 (July 20, 2017)

- Remove `publishr` from prod `dependencies` in npm registry package.

## 0.19.2 (July 20, 2017)

- Switch to `publishr` workflow. (#894, #731, #900)
- Remove `rimraf` from prod `dependencies` in npm registry package.
- Remove `postinstall` from `scripts` in npm registry package. (#794)

## 0.19.1 (May 17, 2017)

### Bug Fixes

- Makes `prop-types` a `dependencies` instead of a `peerDependencies` in `package.json` per the `prop-types` [integration guide](https://github.com/facebook/prop-types#how-to-depend-on-this-package).

## 0.19.0 (May 15, 2017)

### Improvements

- Unreverts PropTypes-related diff, which is now a minor version instead of a patch

## 0.18.4 (May 15, 2017)

### Bug Fixes

- Reverts PropTypes-related diff, which should have been a minor version instead of a patch

## 0.18.3 (May 15, 2017)

### Improvements

- Update dependencies
- Update deprecated React syntax in examples
- Use React "prop-types" package

## 0.18.2 (March 15, 2017)

### Improvements

- Update `inline-style-prefixer` to v2.0.5 (#880).
- Use `React.PureComponent` to avoid unnecessary rendering (#868).
- Update all dependencies

## 0.18.1 (July 19, 2016)

### Bug Fixes

- Fix unused props warning when passing `radiumConfig` to `StyleRoot` (#787).

## 0.18.0 (July 15, 2016)

### Breaking Changes

- Revert `content` auto-quoting in `Style` component (#783).

### Bug Fixes

- Silence React 15.2 unused props deprecation warning (#782).

## 0.17.2 (July 12, 2016)

### Bug Fixes

- Fix `content` values in `Style` component (#719).
- Improve stateless component check to work with native arrow functions (#771).

### Improvements

- Add support for `:disabled` pseudo-class (#689).
- Add plugin to remove nested style objects and prevent `[Object object]` from rendering in the DOM (#703).

## 0.17.1 (March 30, 2016)

### Bug Fixes

- Remove babel modules accidentally published as dependencies.

### Improvements

- Add support for `scopeSelector` without nested selectors in `Style` component.

## 0.17.0 (March 24, 2016)


examples/app.js
/**
 * The examples provided by Formidable Labs are for non-commercial testing and
 * evaluation purposes only. Formidable Labs reserves all rights not expressly
 * granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * FORMIDABLE LABS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable no-use-before-define */

import React from 'react';

import {resetListStyle, resetBoxModel} from './common.styles';
import Button from './components/button';
import ComputedWell from './components/computed-well';
import Radium, {getState, keyframes, Style, StyleRoot} from '../src';

//
// Radium with ES6 class syntax
//
class HoverMessage extends React.Component {
  render() {
    return (
      <div>
        <button key="button" style={{display: 'flex', ':hover': {}}}>
          Hover me!
        </button>
        {getState(this.state, 'button', ':hover') ? (
          <span> Hovering!</span>
        ) : null}
      </div>
    );
  }
}
HoverMessage = Radium(HoverMessage);

//
// Radium with ES7 decorator
//
@Radium
class TwoSquares extends React.Component {
  render() {
    return (
      <div>
        <div key="one" style={[squareStyles.all, squareStyles.one]} />
        <div key="two" style={[squareStyles.all, squareStyles.two]} />
        <div
          disabled
          key="three"
          style={[squareStyles.all, squareStyles.three]}
        />
        <div style={{clear: 'both'}} />
      </div>
    );
  }
}

class Spinner extends React.Component {
  render() {
    return (
      <div>
        <div
          style={[spinnerStyles.inner, {'@media print': {height: '10px'}}]}
        />
      </div>
    );
  }
}

Spinner = Radium(Spinner);

class MultiSpinner extends React.Component {
  render() {
    return (
      <div>
        <div
          style={[
            multiAnimationStyles.inner,
            {'@media print': {height: '10px'}}
          ]}
        />
      </div>
    );
  }
}

MultiSpinner = Radium(MultiSpinner);

const VisitedLink = Radium(() => (
  <a
    href="https://github.com/formidablelabs/radium"
    style={{color: 'gray', ':visited': {color: 'black'}}}
  >
    https://github.com/formidablelabs/radium
  </a>
));

class App extends React.Component {
  _remount() {
    this.setState({shouldRenderNull: true});

    setTimeout(
      function() {
        this.setState({shouldRenderNull: false});
      }.bind(this),
      100
    );
  }

  render() {
    if (this.state && this.state.shouldRenderNull) {
      return null;
    }

    return (
      <StyleRoot>
        <VisitedLink />

        <p />
        <HoverMessage />

        <p />
        <TwoSquares />

        <p />
        <Spinner />

        <p />
        <MultiSpinner />

        <p />
        <Button onClick={this._remount.bind(this)}>Unmount and remount</Button>

        <p />
        <Button>Button</Button>

        <p />
        <Button color="red">Button</Button>

        <p />
        <Button
          style={{
            fontSize: '1.5em',
            borderRadius: 3
          }}
        >
          Button
        </Button>

        <div style={{margin: '20px 0', width: 220}}>
          {Array.apply(null, Array(100)).map(function(_, i) {
            return <div key={'tile' + i} style={tileStyle} />;
          })}
          <div style={{clear: 'both'}} />
        </div>

        <ul style={listStyle}>
          <li>
            Create and use reusable rules with the help of spread operators.
          </li>
        </ul>

        <Style
          rules={{
            body: {
              margin: 0,
              fontFamily: 'Helvetica Neue, Helvetica, Arial, sans-serif'
            },
            mediaQueries: {
              '(max-width: 600px)': {
                body: {
                  background: 'gray'
                }
              },
              '(max-width: 500px)': {
                body: {
                  background: 'blue'
                },
                'p, h1': {
                  color: 'white'
                }
              }
            }
          }}
        />

        <ComputedWell>Click me!</ComputedWell>

        <div className="scoping-class">
          <Style
            rules={{
              span: {
                fontFamily: 'Lucida Console, Monaco, monospace'
              },
              color: 'blue'
            }}
            scopeSelector=".scoping-class"
          />
          <span>This content has scoped styles</span>
        </div>
      </StyleRoot>
    );
  }
}

App = Radium(App);

const squareStyles = {
  all: {
    background: 'black',
    border: 'solid 1px white',
    float: 'left',
    height: 100,
    width: 100
  },
  one: {
    ':hover': {
      background: 'blue'
    }
  },
  two: {
    ':hover': {
      background: 'red'
    }
  },
  three: {
    ':hover': {
      background: 'yellow'
    },
    ':disabled': {
      background: 'red'
    }
  }
};

const tileStyle = {
  display: 'block',
  float: 'left',
  background: '#ccc',
  width: 20,
  height: 20,
  textAlign: 'center',
  border: '1px solid white',
  cursor: 'pointer',

  ':hover': {
    background: '#999'
  }
};

const pulseAnimation = keyframes(
  {
    '0%': {width: '10%'},
    '50%': {width: '50%'},
    '100%': {width: '10%'}
  },
  'pulse'
);

const blendAnimation = Radium.keyframes(
  {
    '0%': {background: 'red'},
    '25%': {background: 'yellow'},
    '50%': {background: 'green'},
    '75%': {background: 'blue'},
    '100%': {background: 'red'}
  },
  'blend'
);

const spinnerStyles = {
  inner: {
    animation: 'x 3s ease 0s infinite',
    animationName: pulseAnimation,
    background: 'blue',
    height: '4px',
    margin: '0 auto'
  }
};

const multiAnimationStyles = {
  inner: {
    animationName: [pulseAnimation, blendAnimation],
    animationDuration: '2.5s, 8s',
    animationIterationCount: 'infinite, infinite',
    animationTimingFunction: 'linear, cubic-bezier(0.1, 0.7, 1.0, 0.1)',
    height: '4px',
    margin: '0 auto'
  }
};

const listStyle = {
  ...resetListStyle,
  ...resetBoxModel,
  marginTop: 15,
  marginBottom: 15,
  marginLeft: 15,
  marginRight: 15
};

export default App;
